validation-test/Reflection/reflect_nested_generic.swift

```diff
@@ -0,0 +1,155 @@
+//X64-NEXT:             (tuple size=16 alignment=8 stride=16 num_extra_inhabitants=0 bitwise_takable=1
+//X64-NEXT:               (field offset=0
+//X64-NEXT:                 (struct size=8 alignment=8 stride=8 num_extra_inhabitants=0 bitwise_takable=1
+//X64-NEXT:                   (field name=_value offset=0
+//X64-NEXT:                     (builtin size=8 alignment=8 stride=8 num_extra_inhabitants=0 bitwise_takable=1))))
+//X64-NEXT:               (field offset=8
+//X64-NEXT:                 (struct size=8 alignment=8 stride=8 num_extra_inhabitants=0 bitwise_takable=1
+//X64-NEXT:                   (field name=_value offset=0
+//X64-NEXT:                     (builtin size=8 alignment=8 stride=8 num_extra_inhabitants=0 bitwise_takable=1))))))
+//X64-NEXT:           (case name=none index=1)))
+//X64-NEXT:       (field name=t offset=17
+//X64-NEXT:         (single_payload_enum size=1 alignment=1 stride=1 num_extra_inhabitants=0 bitwise_takable=1
+//X64-NEXT:           (case name=some index=0 offset=0
+//X64-NEXT:             (tuple size=0 alignment=1 stride=1 num_extra_inhabitants=0 bitwise_takable=1))
+//X64-NEXT:           (case name=none index=1)))))
+//X64-NEXT:   (case name=none index=1))
+
+//TODO:  Work out the layout for 32-bit targets
+
+//CHECK: Mangled name: $s22reflect_nested_generic6Outer1O1CC1SVy_yt_Si_SitGSg
+//CHECK-NEXT: Demangled name: Swift.Optional<reflect_nested_generic.Outer1.C<()>.S<(Swift.Int, Swift.Int)>>
+
+//CHECK: Enum value:
+//CHECK-NEXT: (enum_value name=none index=1)
+
+
+protocol P {}
+
+struct S1: P { var a: Int; var b: Int }
+struct S2: P { }
+
+struct Outer2 {
+  enum E<T: P> {
+    struct Inner {
+      enum F<U: P> {
+        struct Innerer {
+	  var u: U? = nil
+	}
+      case u(U)
+      case a
+      case b
+      }
+    }
+  }
+}
+
+reflect(enum: Outer2.E<S1>.Inner.F<S2>.b)
+
+//CHECK: Reflecting an enum.
+//CHECK-NEXT: Instance pointer in child address space: 0x{{[0-9a-fA-F]+}}
+//CHECK-NEXT: Type reference:
+//CHECK-NEXT: (bound_generic_enum reflect_nested_generic.Outer2.E.Inner.F
+//CHECK-NEXT:   (struct reflect_nested_generic.S2)
+//CHECK-NEXT:   (bound_generic_enum reflect_nested_generic.Outer2.E
+//CHECK-NEXT:     (struct reflect_nested_generic.S1)))
+
+// Note: layout here is same for both 32- and 64-bit platforms
+
+//CHECK: Type info:
+//CHECK-NEXT: (single_payload_enum size=1 alignment=1 stride=1 num_extra_inhabitants=0 bitwise_takable=1
+//CHECK-NEXT:   (case name=u index=0 offset=0
+//CHECK-NEXT:     (struct size=0 alignment=1 stride=1 num_extra_inhabitants=0 bitwise_takable=1))
+//CHECK-NEXT:   (case name=a index=1)
+//CHECK-NEXT:   (case name=b index=2))
+
+//CHECK: Mangled name: $s22reflect_nested_generic6Outer2V1EO1FOy_AA2S1V_AA2S2VG
+//CHECK-NEXT: Demangled name: reflect_nested_generic.Outer2.E<reflect_nested_generic.S1>.F<reflect_nested_generic.S2>
+
+//CHECK: Enum value:
+//CHECK-NEXT: (enum_value name=b index=2)
+
+reflect(enum: Outer2.E<S1>.Inner.F<S2>.Innerer?.none)
```

Comment:
As @augusto2112 suggested, I added a test where the innermost type has no generic type parameters.